### PR TITLE
Re-fix OTA exchange leaks

### LINKED ISSUE #20565: [ota] Exchange leak
#### Problem
When OTA is cancelled due inactivity timeout, the BDX transfer exchange is leaked. Apart from the possibility to run out of exchanges on constrained devices, it causes a Sleepy End Device never getting back to the idle mode, which may drain the battery.
Same thing may occur in different scenarios, such as passing an error to `BDXDownloader::OnPreparedForDownload`.

 #### Proposed Solution
I tried to fix that in a stupid way to close an exchange each time we go back to the idle state, but I wasn't aware of various subtleties of the exchange management and it lead to crashes: https://github.com/project-chip/connectedhomeip/pull/20563.
We need to somehow extend BDX APIs so that we can distinguish cancel events caused by the exchange layer or by external components and close the exchange in the latter cases only.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
#### Problem
OTA exchanges can leak on error.

#### Change overview
1. Re-land (via git cherry-pick) the commit from https://github.com/project-chip/connectedhomeip/pull/20304
2. Fix exchange lifetime management issues in `BDXMessenger` to address the crashes initially observed with #20304.

Fixes https://github.com/project-chip/connectedhomeip/issues/20565

#### Testing
Ran the steps in https://github.com/project-chip/connectedhomeip/pull/20563#issuecomment-1180695605 and verified those do not crash.

### CODE REVIEW

**Comment by mrjerryjohns** [comment]
Is there a unit-test, or YAML test we could add to re-create the failure, and validate the fix?

**Comment by bzbarsky-apple** [comment]
> Is there a unit-test, or YAML test we could add to re-create the failure, and validate the fix?

@Damian-Nordic @isiu-apple @carol-apple

**Comment by carol-apple** [comment]
> > Is there a unit-test, or YAML test we could add to re-create the failure, and validate the fix?
> 
> @Damian-Nordic @isiu-apple @carol-apple

There is one existing test that for a successful transfer. We could build on that but maybe force the provider to be stopped (via SystemCommands) to try to re-create this failure?

**Review by mrjerryjohns — APPROVED** [review]

File: src/app/clusters/ota-requestor/DefaultOTARequestor.cpp [hunk 1]
```diff
@@ -823,10 +825,10 @@ CHIP_ERROR DefaultOTARequestor::StartDownload(OperationalDeviceProxy & devicePro
     Optional<SessionHandle> session = deviceProxy.GetSecureSession();
     VerifyOrReturnError(session.HasValue(), CHIP_ERROR_INCORRECT_STATE);
 
-    mExchangeCtx = exchangeMgr->NewContext(session.Value(), &mBdxMessenger);
-    VerifyOrReturnError(mExchangeCtx != nullptr, CHIP_ERROR_NO_MEMORY);
+    chip::Messaging::ExchangeContext * exchangeCtx = exchangeMgr->NewContext(session.Value(), &mBdxMessenger);
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_NO_MEMORY);
```
Comment (Damian-Nordic) [inline src/app/clusters/ota-requestor/DefaultOTARequestor.cpp:829]: I think we need to close the exchange when one of the functions below returns an error.

**Review by Damian-Nordic — COMMENTED** [review]
I tested this change manually and it doesn't crash anymore but when I tried to inject faults in different places in OTA, I found two more places that need to be fixed:
```diff
diff --git a/src/app/clusters/ota-requestor/BDXDownloader.cpp b/src/app/clusters/ota-requestor/BDXDownloader.cpp
index 8a00a3820..d3b9f0398 100644
--- a/src/app/clusters/ota-requestor/BDXDownloader.cpp
+++ b/src/app/clusters/ota-requestor/BDXDownloader.cpp
@@ -201,10 +201,10 @@ void BDXDownloader::EndDownload(CHIP_ERROR reason)
         {
             mImageProcessor->Abort();
         }
-        SetState(State::kIdle, OTAChangeReasonEnum::kSuccess);
 
         // Because AbortTransfer() will generate a StatusReport to send.
         PollTransferSession();
+        SetState(State::kIdle, OTAChangeReasonEnum::kSuccess);
     }
     else
     {
diff --git a/src/app/clusters/ota-requestor/DefaultOTARequestor.cpp b/src/app/clusters/ota-requestor/DefaultOTARequestor.cpp
index 3f4cf6782..fdc3b4364 100644
--- a/src/app/clusters/ota-requestor/DefaultOTARequestor.cpp
+++ b/src/app/clusters/ota-requestor/DefaultOTARequestor.cpp
@@ -832,8 +832,19 @@ CHIP_ERROR DefaultOTARequestor::StartDownload(OperationalDeviceProxy & devicePro
     mBdxDownloader->SetMessageDelegate(&mBdxMessenger);
     mBdxDownloader->SetStateDelegate(this);
 
-    ReturnErrorOnFailure(mBdxDownloader->SetBDXParams(initOptions, kDownloadTimeoutSec));
-    return mBdxDownloader->BeginPrepareDownload();
+    CHIP_ERROR error = mBdxDownloader->SetBDXParams(initOptions, kDownloadTimeoutSec);
+
+    if (error == CHIP_NO_ERROR)
+    {
+        error = mBdxDownloader->BeginPrepareDownload();
+    }
+
+    if (error != CHIP_NO_ERROR)
+    {
+        mBdxMessenger.Reset();
+    }
+
+    return error;
 }
```

The first is needed because if we call `SetState(State::kIdle, OTAChangeReasonEnum::kSuccess);` to early and close the exchange, we won't manage to send the status report message to inform the provider about the failure. The other one is just another exchange leak if preparation of the download fails in an early stage.

**Review by Damian-Nordic — APPROVED** [review]

**Pull request merged**